File: src/util/crowd_bt.rs
use rgsl::psi::diagamma::psi;
use statrs::function::beta;

pub const GAMMA: f64 = 0.1;
// const LAMBDA: f64 = 1.0; // regularization parameter - not used??
const KAPPA: f64 = 0.0001;
pub const MU_PRIOR: f64 = 0.0;
pub const SIGMA_SQ_PRIOR: f64 = 1.0;
pub const ALPHA_PRIOR: f64 = 10.0;
pub const BETA_PRIOR: f64 = 1.0;
pub const EPSILON: f64 = 0.25;

pub fn update(
    alpha: f64,
    beta: f64,
    mu_winner: f64,
    sigma_sq_winner: f64,
    mu_loser: f64,
    sigma_sq_loser: f64,
) -> (f64, f64, f64, f64, f64, f64) {
    let (updated_alpha, updated_beta, _) = update_annotator(
        alpha,
        beta,
        mu_winner,
        sigma_sq_winner,
        mu_loser,
        sigma_sq_loser,
    );
    let (updated_mu_winner, updated_mu_loser) = update_mus(
        alpha,
        beta,
        mu_winner,
        sigma_sq_winner,
        mu_loser,
        sigma_sq_loser,
    );
    let (updated_sigma_sq_winner, updated_sigma_sq_loser) = update_sigma_sqs(
        alpha,
        beta,
        mu_winner,
        sigma_sq_winner,
        mu_loser,
        sigma_sq_loser,
    );
    return (
        updated_alpha,
        updated_beta,
        updated_mu_winner,
        updated_sigma_sq_winner,
        updated_mu_loser,
        updated_sigma_sq_loser,
    );
}

pub fn expected_information_gain(
    alpha: f64,
    beta: f64,
    mu_a: f64,
    sigma_sq_a: f64,
    mu_b: f64,
    sigma_sq_b: f64,
) -> f64 {
    let (alpha1, beta1, c) = update_annotator(alpha, beta, mu_a, sigma_sq_a, mu_b, sigma_sq_b);
    let (mu_a1, mu_b1) = update_mus(alpha, beta, mu_a, sigma_sq_a, mu_b, sigma_sq_b);
    let (sigma_sq_a1, sigma_sq_b1) =
        update_sigma_sqs(alpha, beta, mu_a, sigma_sq_a, mu_b, sigma_sq_b);
    let prob_a_ranked_above = c;

    let (alpha2, beta2, _) = update_annotator(alpha, beta, mu_b, sigma_sq_b, mu_a, sigma_sq_a);
    let (mu_a2, mu_b2) = update_mus(alpha, beta, mu_b, sigma_sq_b, mu_a, sigma_sq_a);
    let (sigma_sq_a2, sigma_sq_b2) =
        update_sigma_sqs(alpha, beta, mu_b, sigma_sq_b, mu_a, sigma_sq_a);

    return prob_a_ranked_above
        * (divergence_gaussian(mu_a1, sigma_sq_a1, mu_a, sigma_sq_a)
            + divergence_gaussian(mu_b1, sigma_sq_b1, mu_b, sigma_sq_b)
            + GAMMA * divergence_beta(alpha1, beta1, alpha, beta))
        + (1.0 - prob_a_ranked_above)
            * (divergence_gaussian(mu_a2, sigma_sq_a2, mu_a, sigma_sq_a)
                + divergence_gaussian(mu_b2, sigma_sq_b2, mu_b, sigma_sq_b)
                + GAMMA * divergence_beta(alpha2, beta2, alpha, beta));
}

// https://en.wikipedia.org/wiki/Normal_distribution (See Kullback-Leibler divergence)
fn divergence_gaussian(mu1: f64, sigma_sq1: f64, mu2: f64, sigma_sq2: f64) -> f64 {
    let sigma_ratio = sigma_sq1 / sigma_sq2;
    let left_term = f64::powi(mu1 - mu2, 2) / (2.0 * sigma_sq2);
    let right_term = (sigma_ratio - 1.0 - f64::ln(sigma_ratio)) / 2.0;
    return left_term + right_term;
}

// https://en.wikipedia.org/wiki/Beta_distribution (Also Kullback-Leibler divergence)
fn divergence_beta(alpha_1: f64, beta_1: f64, alpha_2: f64, beta_2: f64) -> f64 {
    let ln_term = beta::ln_beta(alpha_2, beta_2) - beta::ln_beta(alpha_1, beta_1);
    let a_term = (alpha_1 - alpha_2) * psi(alpha_1);
    let b_term = (beta_1 - beta_2) * psi(beta_1);
    let ab_term = (alpha_2 - alpha_1 + beta_2 - beta_1) * psi(alpha_1 + beta_1);
    return ln_term + a_term + b_term + ab_term;
}

fn update_annotator(
    alpha: f64,
    beta: f64,
    mu_winner: f64,
    sigma_sq_winner: f64,
    mu_loser: f64,
    sigma_sq_loser: f64,
) -> (f64, f64, f64) {
    let e_mu_winner = f64::exp(mu_winner);
    let e_mu_loser = f64::exp(mu_loser);
    let c1 = e_mu_winner / (e_mu_winner + e_mu_loser)
        + 0.5
            * (sigma_sq_winner + sigma_sq_loser)
            * (e_mu_winner * e_mu_loser * (e_mu_loser - e_mu_winner))
            / f64::powi(e_mu_winner + e_mu_loser, 3);
    let c2 = 1.0 - c1;
    let c = (c1 * alpha + c2 * beta) / (alpha + beta);

    let expected = (c1 * (alpha + 1.0) * alpha + c2 * alpha * beta)
        / (c * (alpha + beta + 1.0) * (alpha + beta));
    let expected_sq = (c1 * (alpha + 2.0) * (alpha + 1.0) * alpha
        + c2 * (alpha + 1.0) * alpha * beta)
        / (c * (alpha + beta + 2.0) * (alpha + beta + 1.0) * (alpha + beta));

    let variance = expected_sq - f64::powi(expected, 2);

    let updated_alpha = (expected - expected_sq) * expected / variance;
    let updated_beta = (expected - expected_sq) * (1.0 - expected) / variance;

    return (updated_alpha, updated_beta, c);
}

fn update_mus(
    alpha: f64,
    beta: f64,
    mu_winner: f64,
    sigma_sq_winner: f64,
    mu_loser: f64,
    sigma_sq_loser: f64,
) -> (f64, f64) {
    let e_mu_winner = f64::exp(mu_winner);
    let e_mu_loser = f64::exp(mu_loser);
    let mult = (alpha * e_mu_winner) / (alpha * e_mu_winner + beta * e_mu_loser)
        - e_mu_winner / (e_mu_winner + e_mu_loser);

    let updated_mu_winner = mu_winner + sigma_sq_winner * mult;
    let updated_mu_loser = mu_loser - sigma_sq_loser * mult;

    return (updated_mu_winner, updated_mu_loser);
}

fn update_sigma_sqs(
    alpha: f64,
    beta: f64,
    mu_winner: f64,
    sigma_sq_winner: f64,
    mu_loser: f64,
    sigma_sq_loser: f64,
) -> (f64, f64) {
    let e_mu_winner = f64::exp(mu_winner);
    let e_mu_loser = f64::exp(mu_loser);
    let mult = (alpha * e_mu_winner * beta * e_mu_loser)
        / f64::powi(alpha * e_mu_winner + beta * e_mu_loser, 2)
        - (e_mu_winner * e_mu_loser / f64::powi(e_mu_winner + e_mu_loser, 2));

    let updated_sigma_sq_winner = sigma_sq_winner * f64::max(1.0 + sigma_sq_winner * mult, KAPPA);
    let updated_sigma_sq_loser = sigma_sq_loser * f64::max(1.0 + sigma_sq_loser * mult, KAPPA);

    return (updated_sigma_sq_winner, updated_sigma_sq_loser);
}

File: server/crowdbt/crowd_bt.go
package crowdbt

import (
	"math"

	"gonum.org/v1/gonum/mathext"
)

const GAMMA = 0.1
const KAPPA = 0.0001
const MU_PRIOR = 0.0
const SIGMA_SQ_PRIOR = 1.0
const ALPHA_PRIOR = 10.0
const BETA_PRIOR = 1.0
const EPSILON = 0.25
const MIN_VIEWS = 4

func Update(alpha float64, beta float64, muWinner float64, sigmaSqWinner float64, muLoser float64, sigmaSqLoser float64) (float64, float64, float64, float64, float64, float64) {
	updatedAlpha, updatedBeta, _ := UpdateAnnotator(alpha, beta, muWinner, sigmaSqWinner, muLoser, sigmaSqLoser)
	updatedMuWinner, updatedMuLoser := UpdateMus(alpha, beta, muWinner, sigmaSqWinner, muLoser, sigmaSqLoser)
	updatedSigmaSqWinner, updatedSigmaSqLoser := UpdateSigmaSqs(alpha, beta, muWinner, sigmaSqWinner, muLoser, sigmaSqLoser)
	return updatedAlpha, updatedBeta, updatedMuWinner, updatedSigmaSqWinner, updatedMuLoser, updatedSigmaSqLoser
}

func ExpectedInformationGain(alpha float64, beta float64, muA float64, sigmaSqA float64, muB float64, sigmaSqB float64) float64 {
	alpha1, beta1, c := UpdateAnnotator(alpha, beta, muA, sigmaSqA, muB, sigmaSqB)
	muA1, muB1 := UpdateMus(alpha, beta, muA, sigmaSqA, muB, sigmaSqB)
	sigmaSqA1, sigmaSqB1 := UpdateSigmaSqs(alpha, beta, muA, sigmaSqA, muB, sigmaSqB)
	probARankedAbove := c

	alpha2, beta2, _ := UpdateAnnotator(alpha, beta, muB, sigmaSqB, muA, sigmaSqA)
	muB2, muA2 := UpdateMus(alpha, beta, muB, sigmaSqB, muA, sigmaSqA)
	sigmaSqB2, sigmaSqA2 := UpdateSigmaSqs(alpha, beta, muB, sigmaSqB, muA, sigmaSqA)

	return probARankedAbove*(DivergenceGaussian(muA1, sigmaSqA1, muA, sigmaSqA)+DivergenceGaussian(muB1, sigmaSqB1, muB, sigmaSqB)+GAMMA*DivergenceBeta(alpha1, beta1, alpha, beta)) + (1.0-probARankedAbove)*DivergenceGaussian(muA2, sigmaSqA2, muA, sigmaSqA) + (DivergenceGaussian(muB2, sigmaSqB2, muB, sigmaSqB) + GAMMA*DivergenceBeta(alpha2, beta2, alpha, beta))
}

// https://en.wikipedia.org/wiki/Normal_distribution (See Kullback-Leibler divergence)
func DivergenceGaussian(mu1 float64, sigmaSq1 float64, mu2 float64, sigmaSq2 float64) float64 {
	sigmaRatio := sigmaSq1 / sigmaSq2
	leftTerm := math.Pow(mu1-mu2, 2) / (2.0 * sigmaSq2)
	rightTerm := (sigmaRatio - 1.0 - math.Log(sigmaRatio)) / 2.0
	return leftTerm + rightTerm
}

// https://en.wikipedia.org/wiki/Beta_distribution (Also Kullback-Leibler divergence)
func DivergenceBeta(alpha1 float64, beta1 float64, alpha2 float64, beta2 float64) float64 {
	lnTerm := mathext.Lbeta(alpha2, beta2) - mathext.Lbeta(alpha1, beta1)
	aTerm := (alpha1 - alpha2) * mathext.Digamma(alpha1)
	bTerm := (beta1 - beta2) * mathext.Digamma(beta1)
	abTerm := (alpha2 - alpha1 + beta2 - beta1) * mathext.Digamma(alpha1+beta1)
	return lnTerm + aTerm + bTerm + abTerm
}

func UpdateAnnotator(alpha float64, beta float64, muWinner float64, sigmaSqWinner float64, muLoser float64, sigmaSqLoser float64) (float64, float64, float64) {
	eMuWinner := math.Exp(muWinner)
	eMuLoser := math.Exp(muLoser)
	c1 := eMuWinner/(eMuWinner+eMuLoser) + 0.5*(sigmaSqWinner+sigmaSqLoser)*(eMuWinner*eMuLoser*(eMuLoser-eMuWinner))/math.Pow(eMuWinner+eMuLoser, 3)
	c2 := 1.0 - c1
	c := (c1*alpha + c2*beta) / (alpha + beta)

	expected := (c1*(alpha+1.0)*alpha + c2*alpha*beta) / (c * (alpha + beta + 1.0) * (alpha + beta))
	expectedSq := (c1*(alpha+2.0)*(alpha+1.0)*alpha + c2*(alpha+1.0)*alpha*beta) / (c * (alpha + beta + 2.0) * (alpha + beta + 1.0) * (alpha + beta))
	variance := expectedSq - math.Pow(expected, 2)

	updatedAlpha := (expected - expectedSq) * expected / variance
	updatedBeta := (expected - expectedSq) * (1.0 - expected) / variance

	return updatedAlpha, updatedBeta, c
}

func UpdateMus(alpha float64, beta float64, muWinner float64, sigmaSqWinner float64, muLoser float64, sigmaSqLoser float64) (float64, float64) {
	eMuWinner := math.Exp(muWinner)
	eMuLoser := math.Exp(muLoser)
	mult := (alpha*eMuWinner)/(alpha*eMuWinner+beta*eMuLoser) - eMuWinner/(eMuWinner+eMuLoser)

	updatedMuWinner := muWinner + mult*sigmaSqWinner
	updatedMuLoser := muLoser - mult*sigmaSqLoser

	return updatedMuWinner, updatedMuLoser
}

func UpdateSigmaSqs(alpha float64, beta float64, muWinner float64, sigmaSqWinner float64, muLoser float64, sigmaSqLoser float64) (float64, float64) {
	eMuWinner := math.Exp(muWinner)
	eMuLoser := math.Exp(muLoser)
	mult := (alpha*eMuWinner*beta*eMuLoser)/math.Pow(alpha*eMuWinner+beta*eMuLoser, 2) - eMuWinner*eMuLoser/math.Pow(eMuWinner+eMuLoser, 2)

	updatedSigmaSqWinner := sigmaSqWinner * math.Max(1.0+mult*sigmaSqWinner, KAPPA)
	updatedSigmaSqLoser := sigmaSqLoser * math.Max(1.0+mult*sigmaSqLoser, KAPPA)

	return updatedSigmaSqWinner, updatedSigmaSqLoser
}
